third-party/realdds/src/dds-notification-server.cpp

```diff
@@ -0,0 +1,138 @@
+                while( _instant_notifications.dequeue( &notification, 1000 ) )
+                {
+                    DDS_API_CALL( _writer->get()->write( &notification ) );
+                }
+                _new_instant_notification = false;
+            }
+        }
+    } )
+{
+    auto topic = topics::device::notification::create_topic( publisher->get_participant(), topic_name.c_str() );
+    _writer = std::make_shared< dds_topic_writer >( topic, publisher );
+
+    _writer->on_publication_matched( [this]( PublicationMatchedStatus const & info ) {
+        if( info.current_count_change == 1 )
+        {
+            LOG_DEBUG( "DataReader " << _writer->topic()->get_participant()->print(
+                           (const dds_guid &)info.last_subscription_handle )
+                                     << " for '" << _writer->topic()->get()->get_name() << "' discovered" );
+            {
+                std::lock_guard< std::mutex > lock( _notification_send_mutex );
+                _send_init_msgs = true;
+            }
+            _send_notification_cv.notify_all();
+        }
+        else if( info.current_count_change == -1 )
+        {
+            LOG_DEBUG( "DataReader " << _writer->topic()->get_participant()->print(
+                           (const dds_guid &)info.last_subscription_handle )
+                                     << " for '" << _writer->topic()->get()->get_name() << "' disappeared" );
+        }
+        else
+        {
+            LOG_ERROR( std::to_string( info.current_count_change )
+                       << " is not a valid value for on_publication_matched" );
+        }
+    } );
+
+    dds_topic_writer::writer_qos wqos( RELIABLE_RELIABILITY_QOS );
+    wqos.history().depth = 10;                                                                      // default is 1
+    wqos.endpoint().history_memory_policy = eprosima::fastrtps::rtps::DYNAMIC_RESERVE_MEMORY_MODE;  // TODO: why?
+    _writer->run( wqos );
+
+    _notifications_loop.start();
+    _active = true;
+}
+
+
+dds_notification_server::~dds_notification_server()
+{
+    // Mark us inactive and wake up the active object so we can properly stop it
+    _active = false;
+    _send_notification_cv.notify_all();
+    _notifications_loop.stop();
+}
+
+
+void dds_notification_server::send_notification( topics::raw::device::notification && notification )
+{
+    std::unique_lock< std::mutex > lock( _notification_send_mutex );
+    if( ! _instant_notifications.enqueue( std::move( notification ) ) )
+    {
+        LOG_ERROR( "error while trying to enqueue a message id:" << notification.id()
+                                                                 << " to instant notifications queue" );
+    }
+};
+
+
+void dds_notification_server::add_init_notification( topics::raw::device::notification && msg )
+{
+    std::unique_lock< std::mutex > lock( _notification_send_mutex );
+    topics::raw::device::notification msg_to_move( msg );
```

Comment:
Why do we need msg_to_move? Can't msg be moved directly?
If msg is not moved, it can be const

Reply:
Might be true, have to look at it.
This is not new code.